confgenerator/config.go

```diff
@@ -72,65 +151,197 @@ func ParseUnifiedConfigAndValidate(input []byte, platform string) (UnifiedConfig
 	return config, nil
 }
 
-type configComponent struct {
-	Type string `yaml:"type" validate:"required"`
+type component interface {
+	Type() string
+	//ValidateType(subagent string, component string, id string, platform string) error
+	//ValidateParameters(subagent string, component string, id string) error
+	//Validate() error
+}
+
+type ConfigComponent struct {
+	ComponentType string `yaml:"type" validate:"required"`
+}
+
+func (c *ConfigComponent) Type() string {
+	return c.ComponentType
+}
+
+type componentType struct {
+	constructor func() component
+	platforms   []string
+}
+
+func (ct componentType) supportsPlatform(ctx context.Context) bool {
+	platform := ctx.Value(platformKey).(string)
+	for _, v := range ct.platforms {
+		if v == platform {
+			return true
+		}
+	}
+	return len(ct.platforms) == 0
+}
+
+type componentTypeRegistry struct {
+	Subagent string
+	Kind     string
+	TypeMap  map[string]*componentType
+}
+
+func (r *componentTypeRegistry) registerType(constructor func() component, platforms ...string) error {
+	name := constructor().(component).Type()
+	if _, ok := r.TypeMap[name]; ok {
+		return fmt.Errorf("Duplicate %s %s type: %q", r.Subagent, r.Kind, name)
+	}
+	if r.TypeMap == nil {
+		r.TypeMap = make(map[string]*componentType)
+	}
+	r.TypeMap[name] = &componentType{constructor, platforms}
+	return nil
+}
+
+var parameterErrRe = regexp.MustCompile(`field (\S+) not found in type \S+`)
+
+func (r *componentTypeRegistry) unmarshalComponentYaml(ctx context.Context, inner *interface{}, unmarshal func(interface{}) error) error {
```

Comment:
// unmarshalComponentYaml is the custom unmarshaller for a component (e.g., the "hostmetrics" receiver, or the "files" receiver).  It first unmarshals into a struct containing only the "type" field, then looks up the struct with the full set of fields for that type, and finally unmarshals into an instance of that struct.